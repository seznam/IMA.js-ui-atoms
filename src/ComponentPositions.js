import { UserAgent } from '@ima/plugin-useragent';

/**
 * A bounding client rectangle.
 * @typedef {{
 *            top: number,
 *            left: number,
 *            width: number,
 *            height: number
 *          }} BoundingClientRect
 */

/**
 * Component positions helper.
 */
export default class ComponentPositions {
  static get $dependencies() {
    return ['$Window', UserAgent];
  }

  /**
   * Initializes the helper.
   *
   * @param {ima.window.Window} window
   * @param {UserAgent} userAgent
   */
  constructor(window, userAgent) {
    /**
     * @type {ima.window.Window}
     */
    this._window = window;

    /**
     * @type {UserAgent}
     */
    this._userAgent = userAgent;
  }

  /**
   * Convert string to number.
   *
   * @param {string} string
   * @return {number}
   */
  convertToNumber(string) {
    let number = 0;

    try {
      number = parseInt(string, 10);
    } catch (e) {
      number = 0;
    }

    return number;
  }

  /**
   * Returns percent of visibility defined area in window viewport.
   *
   * @param {BoundingClientRect} elmRect
   * @return {number} The percent of visibility.
   */
  getPercentOfVisibility(elmRect) {
    if (!elmRect) {
      throw new Error(
        `Element rect is required. Call getBoundingClientRect()` +
          ` method on element or give object with properties { top: number,` +
          ` left: number, width: number, height: number }.`
      );
    }

    if (!this._window.isClient()) {
      return 0;
    }

    let windowViewportRect = this.getWindowViewportRect();

    let intersectionRect = this.getRectsIntersection(
      windowViewportRect,
      elmRect
    );
    let percent =
      ((intersectionRect.width * intersectionRect.height) /
        (elmRect.width * elmRect.height)) *
      100;

    return isNaN(percent) ? 0 : percent;
  }

  /**
   * Returns an intersection rectangle of two defined reactangles.
   *
   * @param {BoundingClientRect} rect1
   * @param {BoundingClientRect} rect2
   * @return {BoundingClientRect} The intersection rectangle.
   */
  getRectsIntersection(rect1, rect2) {
    let top = this.getNumberFromRange(rect2.top, rect1.top, rect1.height);
    let left = this.getNumberFromRange(rect2.left, rect1.left, rect1.width);
    let bottom = this.getNumberFromRange(
      rect2.top + rect2.height,
      rect1.top,
      rect1.height
    );
    let right = this.getNumberFromRange(
      rect2.left + rect2.width,
      rect1.left,
      rect1.width
    );
    let width = right - left;
    let height = bottom - top;

    return { top, left, width, height };
  }

  /**
   * Returns number from defined range, if number is not in defined range return
   * min or max depends on number.
   *
   * @param {number} number
   * @param {number} min
   * @param {number} max
   * @return {number}
   */
  getNumberFromRange(number, min, max) {
    return Math.min(Math.max(number, min), max);
  }

  /**
   * Returns window viewport rect.
   *
   * @return {BoundingClientRect}
   */
  getWindowViewportRect() {
    let win = this._window.getWindow();
    let top = 0;
    let left = 0;
    let width = 0;
    let height = 0;

    if (this._window.isClient()) {
      width = win.innerWidth;
      height = win.innerHeight;
    }

    return { top, left, width, height };
  }

  /**
   * Returns window scroll position.
   *
   * @return {{top: number, left: number}}
   */
  getWindowScrollPosition() {
    let left = window.pageXOffset ? window.pageXOffset : 0;
    let top = window.pageYOffset ? window.pageYOffset : 0;

    return { top, left };
  }

  /**
   * Returns the size of an element and its position relative to the viewport
   * and add extended value to returned rect.
   *
   * @param {Element} element
   * @param {{width: number, height: number}} size
   * @param {number} extended
   * @return {BoundingClientRect}
   */
  getBoundingClientRect(
    element,
    { width, height } = { width: 0, height: 0 },
    extended = 0
  ) {
    if (!element || typeof element.getBoundingClientRect !== 'function') {
      throw new Error(
        `Element rect is required with callable getBoundingClientRect()` +
          ` method on element.`
      );
    }

    const clientRect = element.getBoundingClientRect();
    const isInvisibleElement =
      clientRect.top === 0 &&
      clientRect.left === 0 &&
      clientRect.width === 0 &&
      clientRect.height === 0;

    if (isInvisibleElement) {
      const { top, left, width, height } = clientRect;

      return { top, left, width, height };
    }

    let elmRectStyle = {
      top: clientRect.top - extended,
      left: clientRect.left - extended,
      width: clientRect.width + 2 * extended,
      height: (clientRect.height || height || 0 / width || 0) + 2 * extended,
    };

    return this._fixBoundingClientRectOnIOS(elmRectStyle);
  }

  /**
   * Applies a fix for iOS 8+ bug, where overscroll messes up
   * getBoundingClientRect()'s top value on all iOS webkit based devices:
   * https://github.com/lionheart/openradar-mirror/issues/6233
   *
   * @param {BoundingClientRect} boundingClientRect A bounding client rectangle.
   * @return {BoundingClientRect} A fixed bounding client rectangle.
   */
  _fixBoundingClientRectOnIOS(boundingClientRect) {
    if (this._userAgent.getOSFamily() !== 'iOS') {
      return boundingClientRect;
    }

    const window = this._window.getWindow();
    const maxScrollHeight = this._getMaxScrollHeight();
    const rect = Object.assign({}, boundingClientRect);

    if (window.scrollY < 0 && rect.top > 0) {
      rect.top += window.scrollY;
    } else if (window.scrollY > maxScrollHeight && rect.top < 0) {
      rect.top += window.scrollY - maxScrollHeight;
    }

    return rect;
  }

  /**
   * Returns maximum available scroll height (minus viewport height).
   *
   * @returns {number}
   */
  _getMaxScrollHeight() {
    const window = this._window.getWindow();
    const document = this._window.getDocument();

    return (
      Math.max(
        document.body.scrollHeight || 0,
        document.body.offsetHeight || 0,
        document.documentElement.clientHeight || 0,
        document.documentElement.offsetHeight || 0,
        document.documentElement.scrollHeight || 0
      ) - window.innerHeight
    );
  }
}
