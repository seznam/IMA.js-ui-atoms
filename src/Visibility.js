/**
 * Visibility helper.
 */
const DEFAULT_VISIBILITY_INTERVAL = 180;
const DEFAULT_THROTTLE_INTERVAL = 60;

export default class Visibility {

	static get $dependencies() {
		return [
			'$Window'
		];
	}

	constructor(window) {
		/**
		 * @property _window
		 * @type {ima.window.Window}
		 */
		this._window = window;

		/**
		 * @property _entries
		 * @type {Map}
		 */
		this._entries = new Map();

		/**
	   * @property _nextLoopEntriesMaxInterval
	   * @type {number}
	   */
			this._nextLoopEntriesMaxInterval = 0;

		/**
	   * @property _nextLoopEntriesTimeout
	   * @type {timer ID|null}
	   */
			this._nextLoopEntriesTimeout = null;

		/**
		 * @property _uniquiId
		 * @type {number}
		 */
		this._uniquiId = 0;

		/**
		 * @property _runningLoop
		 * @type {boolean}
		 */
		this._runningLoop = false;

		/**
		 * @property _throttledVisibilityLoop
		 * @type {function}
		 */
		this._throttledVisibilityLoop = this.throttle(
			this._visibilityLoop,
			DEFAULT_THROTTLE_INTERVAL,
			this
		);
	}

	/**
	 * Register handlers to visibility loop
	 *
	 * @param {function} reader
	 * @param {function} writer
	 * @param {{ visibilityInterval: number }} options
	 * @return {number} The registered id
	 */
	register(reader, writer, options = { visibilityInterval: DEFAULT_VISIBILITY_INTERVAL }) {
		let id = this._generateId();

		this._entries.set(
			id,
			{ reader, writer, options, lastLoop: Date.now() - options.visibilityInterval - 1 }
		);

		this._listenOnEvents();
		this._throttledVisibilityLoop();

		return id;
	}

	/**
	 * Unregister handlers from visibility loop
	 *
	 * @param {number} The registered id
	 */
	unregister(id) {
		this._entries.delete(id);

		if (this._entries.size === 0) {
			this._unlistenOnEvents();
		}
	}

	/**
	 * It cut down calling the event handler for defined interval. The throttle
	 * method use requestAnimationFrame function which is called during page
	 * scrolling.
	 *
	 * @method throttle
	 * @param {function(...)} eventHandler
	 * @param {number?} interval
	 * @param {Object?} context
	 * @return {function(...)} The throttled event
	 */
	throttle(eventHandler, interval, context) {
		let win = this._window.getWindow();
		interval = interval || 0;
		let callTime = 0;
		let lastArguments = null;
		
		clearTimeout(this._nextLoopEntriesTimeout);

		if (context) {
			eventHandler = eventHandler.bind(context);
		}

		if (!this._window.isClient()) {
			return eventHandler;
		}

		function suspendAction() {
			if (callTime <= Date.now() || !win.requestAnimationFrame) {
				callTime = 0;
				eventHandler(...lastArguments);
			} else {
				win.requestAnimationFrame(suspendAction);
			}
		}

		return function throttle(...rest) {
			lastArguments = rest;

			if (!callTime) {
				callTime = Date.now() + interval;
				suspendAction();
			}
		};
	}

	/**
	 * The main visibility loop is constructed for purpose performance improvements.
	 * The values are read at first and then the writer method is called with previous
	 * recorded value.
	 */
	_visibilityLoop() {
		let lastLoopRunning = Date.now();
		
		this._nextLoopEntriesMaxInterval = 0;
		clearTimeout(this._nextLoopEntriesTimeout);

		Array.from(this._entries.values()).filter((loopEntry) => {
			if ((loopEntry.lastLoop + loopEntry.options.visibilityInterval) <= Date.now()) {
				return true;
			} else {
				this._nextLoopEntriesMaxInterval = Math.max(this._nextLoopEntriesMaxInterval, loopEntry.options.visibilityInterval, DEFAULT_THROTTLE_INTERVAL + 1);
				return false;
			}
		}).map((loopEntry) => {
			loopEntry.lastLoop = lastLoopRunning;
			let readedValue = typeof loopEntry.reader === 'function' ? loopEntry.reader() : null;

			return {
				readedValue,
				reader: loopEntry.reader,
				writer: loopEntry.writer
			};
		}).map((loopEntry) => {
			if (typeof loopEntry.writer === 'function') {
				loopEntry.writer(loopEntry.readedValue);
			}
		});

		// Fires new loop in the end of the throttle if any entry.writer has left unhandled
		if (this._nextLoopEntriesMaxInterval) {
			const _this = this;
			this._nextLoopEntriesTimeout = setTimeout( () => { _this._visibilityLoop(); }, this._nextLoopEntriesMaxInterval + 1 );
		}
	}

	/**
	 * Generate unique id.
	 *
	 * @return {number}
	 */
	_generateId() {
		this._uniquiId += 1;

		return this._uniquiId;
	}

	/**
	 * The visibility helper start checking visibility of registered entries.
	 */
	_listenOnEvents() {
		if (this._runningLoop) {
			return;
		}

		this._runningLoop = true;
		this._window.bindEventListener(this._window.getWindow(), 'resize', this._throttledVisibilityLoop);
		this._window.bindEventListener(this._window.getWindow(), 'scroll', this._throttledVisibilityLoop);
	}

	/**
	 * The visibility helper stop checking visibility of registered entries.
	 */
	_unlistenOnEvents() {
		if (!this._runningLoop) {
			return;
		}

		this._runningLoop = false;
		this._window.unbindEventListener(this._window.getWindow(), 'resize', this._throttledVisibilityLoop);
		this._window.unbindEventListener(this._window.getWindow(), 'scroll', this._throttledVisibilityLoop);
	}

}
