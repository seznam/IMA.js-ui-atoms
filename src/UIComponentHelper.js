import ComponentPositions from './ComponentPositions';
import Visibility from './Visibility';

/**
 * UI component helper.
 */
export default class UIComponentHelper {

	static get $dependencies() {
		return [
			'$Router',
			ComponentPositions,
			Visibility,
			'$CssClasses'
		];
	}

	/**
	 * Initializes the helper.
	 *
	 * @param {ima.router.Router} router
	 * @param {ComponentPositions} componentPositions
	 * @param {Visibility} visibility
	 * @param {function(...?(boolean|string|React.Component|Object<string, boolean>)): string} cssClassNameProcessor
	 */
	constructor(router, componentPositions, visibility, cssClassNameProcessor) {
		/**
		 * IMA Router
		 *
		 * @type {ima.router.Router}
		 */
		this._router = router;

		/**
		 * Component position
		 *
		 * @type {ComponentPosition}
		 */
		this._componentPositions = componentPositions;

		/**
		 * Visibility helper
		 *
		 * @type {Visibility}
		 */
		this._visibility = visibility;

		/**
		 * @type {function(...?(boolean|string|React.Component|Object<string, boolean>)): string}
		 */
		this._cssClassNameProcessor = cssClassNameProcessor;
	}

	/**
	 * The public getter for visibility helper.
	 *
	 * @return {Visibility}
	 */
	get visibility() {
		return this._visibility;
	}

	/**
	 * The public getter for component positions helper.
	 *
	 * @return {ComponentPositions}
	 */
	get componentPositions() {
		return this._componentPositions;
	}

	/**
	 * Returns true if page may be rendered as amp page.
	 *
	 * @return {boolean}
	 */
	isAmp() {
		let ampParam = null;

		try {
			ampParam = this._router.getCurrentRouteInfo().params.amp;
		} catch (error) {
			ampParam = false;
		}

		return ampParam === true || ampParam === '1';
	}

	/**
	 * Filters the provided properties and returns only the properties which's
	 * names start with the {@code data-} prefix.
	 *
	 * @param {Object<string, *>} props
	 * @return {Object<string, (number|string)>}
	 */
	getDataProps(props) {
		let dataProps = {};

		for (let propertyName of Object.keys(props)) {
			if (/^data-/.test(propertyName)) {
				dataProps[propertyName] = props[propertyName];
			}
		}

		return dataProps;
	}

	/**
	 * Filters the provided properties and returns only the properties which's
	 * names start with the {@code aria-} prefix.
	 *
	 * @param {Object<string, *>} props
	 * @return {Object<string, (number|string)>}
	 */
	getAriaProps(props) {
		let ariaProps = {};

		for (let propertyName of Object.keys(props)) {
			if (/^aria-/.test(propertyName)) {
				ariaProps[propertyName] = props[propertyName];
			}
		}

		return ariaProps;
	}

	/**
	 * Generate a string of CSS classes from the properties of the passed-in
	 * object that resolve to true.
	 *
	 * @param {...?(string|Object<string, boolean>)} classRuleGroups CSS
	 *        classes in a string separated by whitespace, or a map of CSS
	 *        class names to boolean values. The CSS class name will be
	 *        included in the result only if the value is {@code true}.
	 *        Declarations in the later class rule group will override the
	 *        declarations in the previous group.
	 * @return {string} String of CSS classes that had their property resolved
	 *         to {@code true}.
	 */
	cssClasses(...classRuleGroups) {
		return this._cssClassNameProcessor(...classRuleGroups);
	}

	/**
	 * Only Facade mehtod of ComponentPositions.convertToNumber function
	 * for backward compatibility.
	 *
	 * @deprecated removed with next version
	 * @param {string} string
	 * @return {number}
	 */
	convertToNumber(string) {
		return this._componentPositions.convertToNumber(string);
	}

	/**
	 * Only Facade mehtod of ComponentPositions.getWindowViewportRect function
	 * for backward compatibility.
	 *
	 * @deprecated removed with next version
	 * @return {{top: number, left: number, width: number, height: number}}
	 */
	getWindowViewportRect() {
		return this._componentPositions.getWindowViewportRect();
	}

	/**
	 * Only Facade mehtod of ComponentPositions.getBoundingClientRect function
	 * for backward compatibility.
	 *
	 * @deprecated removed with next version
	 * @param {Element} element
	 * @param {{width: number, height: number}} size
	 * @param {number} extended
	 * @return {{top: number, left: number, width: number, height: number}}
	 */
	getBoundingClientRect(element, size, extended) {
		return this._componentPositions.getBoundingClientRect(element, size, extended);
	}

	/**
	 * Only Facade mehtod of ComponentPositions.getPercentOfVisibility function
	 * for backward compatibility.
	 *
	 * @deprecated removed with next version
	 * @param {{top: number, left: number, width: number, height: number}} elmRect
	 * @return {number} The percent of visibility.
	 */
	getPercentOfVisibility(elmRect) {
		return this._componentPositions.getPercentOfVisibility(elmRect);
	}

	/**
	 * @param {HTMLElement} element
	 * @param {{ width: ?number, height: ?number, extendedPadding: ?number }} options
	 * @return {function}
	 */
	getVisibilityReader(element, options) {
		let self = this;

		return function readVisibility() {
			let elementRect = self._componentPositions.getBoundingClientRect(
				element,
				{ width: options.width, height: options.height },
				options.extendedPadding
			);

			return self._componentPositions.getPercentOfVisibility(elementRect);
		};
	}

	/**
	 * Only Facade mehtod of Visibility.throttle function
	 * for backward compatibility.
	 *
	 * @deprecated removed with next version
	 * @param {function(...)} eventHandler
	 * @param {number?} interval
	 * @param {Object?} context
	 * @return {function(...)} The throttled event
	 */
	throttle(eventHandler, interval, context) {
		return this._visibility.throttle(eventHandler, interval, context);
	}

	/**
	 * Register component to main visibility loop
	 *
	 * @param {function} reader
	 * @param {function} writer
	 * @param {{ visibilityInterval: number }} options
	 * @return {number} The registered id
	 */
	registerComponentToVisibility(reader, writer, options = { visibilityInterval: 180 }) {
		return this._visibility.register(reader, writer, options);
	}

	/**
	 * Unregister component from main visibility loop
	 *
	 * @param {number} visibilityId
	 */
	unregisterComponentToVisibility(visibilityId) {
		this._visibility.unregister(visibilityId);
	}
}
