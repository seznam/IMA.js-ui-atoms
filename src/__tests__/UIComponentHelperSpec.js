import { toMockedInstance } from 'to-mock';
import classnames from 'classnames';

import ComponentPositions from '../ComponentPositions';
import UIComponentHelper from '../UIComponentHelper';
import Visibility from '../Visibility';
import { Infinite } from 'infinite-circle';

import _router from '../mocks/router';
import _window from '../mocks/window';

describe('UIComponentHelper', () => {
  let uiComponentHelper = null;
  let visibility = toMockedInstance(Visibility);
  let componentPositions = toMockedInstance(ComponentPositions);
  let infinite = toMockedInstance(Infinite);

  beforeEach(() => {
    uiComponentHelper = new UIComponentHelper(
      _router,
      _window,
      componentPositions,
      visibility,
      infinite,
      classnames
    );
  });

  describe('isAmp method', () => {
    it('should return true if url query contains amp flag of value true', () => {
      spyOn(_router, 'getCurrentRouteInfo').and.returnValue({
        params: { amp: true },
      });

      expect(uiComponentHelper.isAmp()).toBeTruthy();
    });

    it('should return true if url query contains amp flag of value "1"', () => {
      spyOn(_router, 'getCurrentRouteInfo').and.returnValue({
        params: { amp: '1' },
      });

      expect(uiComponentHelper.isAmp()).toBeTruthy();
    });

    it('should return false if url query does not contain amp flag', () => {
      spyOn(_router, 'getCurrentRouteInfo').and.returnValue({
        params: { amp: '0' },
      });

      expect(uiComponentHelper.isAmp()).toBeFalsy();
    });

    it('should return false if url query not contains amp flag', () => {
      expect(uiComponentHelper.isAmp()).toBeFalsy();
    });
  });

  describe('getDataProps method', () => {
    let dataProps = {
      'data-e2e': 'something',
      'data-key': 'key',
    };
    let props = Object.assign({ key: 'key' }, dataProps);

    it('should return only attributes with name data-*', () => {
      expect(uiComponentHelper.getDataProps(props)).toEqual(dataProps);
    });
  });

  describe('getAriaProps method', () => {
    let ariaProps = {
      'aria-label': 'something',
      'aria-hidden': true,
    };
    let props = Object.assign({ key: 'key' }, ariaProps);

    it('should return only attributes with name aria-*', () => {
      expect(uiComponentHelper.getAriaProps(props)).toEqual(ariaProps);
    });
  });

  describe('getEventProps method', () => {
    let eventProps = {
      onKeyDown: () => {},
      onClick: function () {},
    };
    let props = Object.assign(
      {
        key: 'key',
        once: 'notValid',
        oNsubmit: 'notValid',
        onSubmit: 'notValid',
        onDrag: {},
      },
      eventProps
    );

    it('should return only attributes with name on[A-Z]*', () => {
      expect(uiComponentHelper.getEventProps(props)).toEqual(eventProps);
    });
  });

  describe('serializeObjectToNoScript method', () => {
    let ariaProps = {
      'aria-label': 'something',
      'aria-hidden': true,
    };

    it('should return serialized object to string for noscript tag', () => {
      expect(
        uiComponentHelper.serializeObjectToNoScript(ariaProps)
      ).toMatchSnapshot();
    });
  });

  describe('cssClasses method', () => {
    it('should compose CSS class names', () => {
      expect(
        uiComponentHelper.cssClasses(
          'stuff another-foo',
          {
            foo: true,
            bar: false,
            another: false,
            'more-things': true,
          },
          'things'
        )
      ).toBe('stuff another-foo foo more-things things');
    });
  });

  describe('getVisibilityReader method', () => {
    it('return base visibility reader function', () => {
      expect(
        typeof uiComponentHelper.getVisibilityReader({}, {}) === 'function'
      ).toBeTruthy();
    });
  });

  describe('wrapVisibilityWriter method', () => {
    function writer(...args) {
      return `writer value: ${args.join(',')}`;
    }

    function wrapWriterTest(payload) {
      const entry = { payload };
      const parse = uiComponentHelper.wrapVisibilityWriter(writer);

      return parse(entry);
    }

    const OBSERVER = {};
    const PAYLOAD = 42;
    const PAYLOAD_OBJECT = {
      observer: OBSERVER,
      intersectionObserverEntry: {
        intersectionRatio: 0.3,
        isIntersecting: true,
      },
    };
    const NONINTERSECTED_PAYLOAD_OBJECT = {
      observer: OBSERVER,
      intersectionObserverEntry: {
        intersectionRatio: 0,
        isIntersecting: false,
      },
    };
    const BUGGED_PAYLOAD_OBJECT = {
      observer: OBSERVER,
      intersectionObserverEntry: {
        intersectionRatio: 0,
        isIntersecting: true,
      },
    };

    it('should return a return value of writer(PAYLOAD)', () => {
      const result = wrapWriterTest(PAYLOAD);
      expect(result).toBe(writer(PAYLOAD));
    });

    it('should return a return value of writer(30, OBSERVER)', () => {
      const result = wrapWriterTest(PAYLOAD_OBJECT);
      expect(result).toBe(writer(30, OBSERVER));
    });

    it('should return a return value of writer(0, OBSERVER)', () => {
      const result = wrapWriterTest(NONINTERSECTED_PAYLOAD_OBJECT);
      expect(result).toBe(writer(0, OBSERVER));
    });

    it('should return a return value of writer(100, OBSERVER)', () => {
      const result = wrapWriterTest(BUGGED_PAYLOAD_OBJECT);
      expect(result).toBe(writer(100, OBSERVER));
    });
  });
});
