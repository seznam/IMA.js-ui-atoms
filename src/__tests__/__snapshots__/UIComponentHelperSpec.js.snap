// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UIComponentHelper serializeObjectToNoScript method should return serialized object to string for noscript tag 1`] = `" aria-label=\\"something\\" aria-hidden=\\"true\\""`;
